src/OpenSSL/SSL.py

```diff
@@ -161,6 +164,47 @@ class SysCallError(Error):
     pass
 
 
+def _make_buffer_checker(cffi_version_string):
+    """
+    Returns a callable that converts a sendable buffer object into something
+    that can be sent by OpenSSL using CFFI. Has different supported types
+    depending on the CFFI version.
+
+    :param cffi_version_string: The version of CFFI in use, from
+        ffi.__version__.
+    """
+    cffi_version = parse(cffi_version_string)
+    if cffi_version >= Version("1.7"):
+        allowed_types = (bytes, bytearray)
+        error_message = (
+            "data must be a memoryview, buffer, bytearray or byte string"
+        )
+        make_buffer = _ffi.from_buffer
+    else:
+        allowed_types = (bytes,)
+        error_message = "data must be a memoryview, buffer, or byte string"
+        make_buffer = partial(_ffi.new, "char[]")
+
+    def buffer_factory(buf):
+        """
+        Given a sendable data buffer passed into send or sendall, returns a
+        CFFI object representing it.
+
+        :param buf: The sendable data buffer the user wishes to send.
+        """
+        if isinstance(buf, _memoryview):
+            buf = buf.tobytes()
+        if isinstance(buf, _buffer):
+            buf = str(buf)
```

Comment:
1) It may not matter much in practice, but I think on recent CFFI these lines are unnecessary and are inefficient in that they impose an extra copy? 
2) I might add a comment on the `str` line noting that this branch can only be taken on py2 (because py3 has no buffer objects)